File: php/Day2.php
<?php
require_once(__DIR__ . '/Day.php');

class Day2 extends Day
{
    protected function part1($puzzleInput)
    {
        $depth = 0;
        $horizontalPosition = 0;
        
        foreach ($puzzleInput as $line) {
            $split = explode(" ", $line);    
            $units = (int) $split[1];
            $direction = trim($split[0]);

            if ($direction === "forward") {
                $horizontalPosition += $units;
            } else if ($direction === "up") {
                $depth -= $units;
            } else if ($direction === "down") {
                $depth += $units;
            }
        }

        return $depth * $horizontalPosition;
    }

    protected function part2($puzzleInput)
    {
        $depth = 0;
        $horizontalPosition = 0;
        $aim = 0;

        foreach ($puzzleInput as $line) {
            $split = explode(" ", $line);
            $units = (int) $split[1];
            $direction = trim($split[0]);

            if ($direction === "forward") {
                $horizontalPosition += $units;
                $depth += $units * $aim;
            } else if ($direction === "up") {
                $aim -= $units;
            } else if ($direction === "down") {
                $aim += $units;
            }
        };

        return $depth * $horizontalPosition;
    }
}

$day = new Day2();
$day->run(2);

File: c#/AdventOfCode/AdventOfCode/Day2.cs
namespace AdventOfCode;

public class Day2 : Day
{
    private enum CommandType
    {
        Forward,
        Down,
        Up
    }
    
    private class Command
    {
        public readonly int Units;
        public readonly CommandType Type;

        public Command(string row)
        {
            var splits = row.Split(' ');
            Units = int.Parse(splits[1]);
            Type = splits[0] switch
            {
                "up" => CommandType.Up,
                "down" => CommandType.Down,
                "forward" => CommandType.Forward,
                _ => throw new ArgumentOutOfRangeException(nameof(row), row, "Invalid command")
            };
        }
    }
    
    public override int Part1(string[] puzzleInput)
    {
        var x = 0;
        var y = 0;
        
        foreach (var row in puzzleInput)
        {
            var command = new Command(row);
            switch (command.Type)
            {
                case CommandType.Up:
                    y -= command.Units;
                    break;
                case CommandType.Down:
                    y += command.Units;
                    break;
                case CommandType.Forward:
                    x += command.Units;
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        return x * y;
    }

    public override int Part2(string[] puzzleInput)
    {
        var horizontalPosition = 0;
        var depth = 0;
        var aim = 0;
        
        foreach (var row in puzzleInput)
        {
            var command = new Command(row);
            switch (command.Type)
            {
                case CommandType.Up:
                    aim -= command.Units;
                    break;
                case CommandType.Down:
                    aim += command.Units;
                    break;
                case CommandType.Forward:
                    horizontalPosition += command.Units;
                    depth += aim * command.Units;
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        return horizontalPosition * depth;
    }
}